object-api-putobject_test.go

```diff
@@ -0,0 +1,213 @@
+	bucket := "minio-bucket"
+	object := "minio-object"
+
+	// Create bucket.
+	err := obj.MakeBucket(bucket)
+	if err != nil {
+		// Failed to create newbucket, abort.
+		t.Fatalf("%s : %s", instanceType, err.Error())
+	}
+
+	// Creating a dummy bucket for tests.
+	err = obj.MakeBucket("unused-bucket")
+	if err != nil {
+		// Failed to create newbucket, abort.
+		t.Fatalf("%s : %s", instanceType, err.Error())
+	}
+
+	// Take 6 disks down, one more we loose quorum on 16 disk node.
+	for _, disk := range disks[:6] {
+		removeAll(disk)
+	}
+
+	testCases := []struct {
+		bucketName      string
+		objName         string
+		inputReaderData string
+		inputMeta       map[string]string
+		intputDataSize  int64
+		// flag indicating whether the test should pass.
+		shouldPass bool
+		// expected error output.
+		expectedMd5   string
+		expectedError error
+	}{
+		// Validating for success cases.
+		{bucket, object, "abcd", map[string]string{"md5Sum": "e2fc714c4727ee9395f324cd2e7f331f"}, int64(len("abcd")), true, "", nil},
+		{bucket, object, "efgh", map[string]string{"md5Sum": "1f7690ebdd9b4caf8fab49ca1757bf27"}, int64(len("efgh")), true, "", nil},
+		{bucket, object, "ijkl", map[string]string{"md5Sum": "09a0877d04abf8759f99adec02baf579"}, int64(len("ijkl")), true, "", nil},
+		{bucket, object, "mnop", map[string]string{"md5Sum": "e132e96a5ddad6da8b07bba6f6131fef"}, int64(len("mnop")), true, "", nil},
+		{bucket, object, "mnop", map[string]string{"md5Sum": "e132e96a5ddad6da8b07bba6f6131fef"}, int64(len("mnop")), false, "", InsufficientWriteQuorum{}},
+	}
+
+	for i, testCase := range testCases[:3] {
+		actualMd5Hex, actualErr := obj.PutObject(testCase.bucketName, testCase.objName, testCase.intputDataSize, bytes.NewBufferString(testCase.inputReaderData), testCase.inputMeta)
+		if actualErr != nil && testCase.shouldPass {
+			t.Errorf("Test %d: %s: Expected to pass, but failed with: <ERROR> %s.", i+1, instanceType, actualErr.Error())
+		}
+		if actualErr == nil && !testCase.shouldPass {
+			t.Errorf("Test %d: %s: Expected to fail with <ERROR> \"%s\", but passed instead.", i+1, instanceType, testCase.expectedError.Error())
+		}
+		// Failed as expected, but does it fail for the expected reason.
+		if actualErr != nil && !testCase.shouldPass {
+			if testCase.expectedError.Error() != actualErr.Error() {
+				t.Errorf("Test %d: %s: Expected to fail with error \"%s\", but instead failed with error \"%s\" instead.", i+1,
+					instanceType, testCase.expectedError.Error(), actualErr.Error())
+			}
+		}
+		// Test passes as expected, but the output values are verified for correctness here.
+		if actualErr == nil && testCase.shouldPass {
+			// Asserting whether the md5 output is correct.
+			if testCase.inputMeta["md5Sum"] != actualMd5Hex {
+				t.Errorf("Test %d: %s: Calculated Md5 different from the actual one %s.", i+1, instanceType, actualMd5Hex)
+			}
+		}
+	}
+
+	// This causes quorum failure verify.
+	removeAll(disks[len(disks)-1])
```

Comment:
> Wouldn't it be simpler to move the last test case from the testcases slice, since we perform different checks instead of restting index to point to the last element in the slice?

Done